src/dashboard/src/contrib/mcp/client.py

```diff
@@ -27,25 +27,94 @@
 LOGGER = logging.getLogger('archivematica.dashboard.mcp.client')
 
 
-class RPCError(Exception):
-    pass
+class RPCGearmanClientError(Exception):
+    """Base exception."""
 
 
-class NoJobFoundError(Exception):
-    def __init__(self, *args, **kwargs):
-        try:
-            message = args[0]
-        except IndexError:
+class RPCError(RPCGearmanClientError):
+    """Unexpected error."""
+
+
+class RPCServerError(RPCGearmanClientError):
+    """Server application errors.
+
+    When the worker processes the job successfully but the response includes
+    an error.
+    """
+    GENERIC_ERROR_MSG = "The server failed to process the request"
+
+    def __init__(self, payload=None):
+        super(RPCServerError, self).__init__(self._process_error(payload))
+
+    def _process_error(self, payload):
+        """Extracts the error message from the payload."""
+        if payload is None or not isinstance(payload, dict):
+            return self.GENERIC_ERROR_MSG
+        message = payload.get("message", "Unknown error message")
+        handler = payload.get("function")
+        if handler:
+            message += " [handler=%s]" % (handler,)
+        return message
+
+
+class TimeoutError(RPCGearmanClientError):
+    """Deadline exceeded.
+
+    >> response = client.submit_job(
+           "doSomething", cPickle.dumps(data),
+           background=False, wait_until_complete=True,
+           poll_timeout=INFLIGHT_POLL_TIMEOUT)
+       if response.state == gearman.JOB_CREATED:
+           raise TimeoutError()
+
+    At this point we give up and raise this exception.
+    """
+    def __init__(self, timeout=None):
+        message = "Deadline exceeded"
+        if timeout is not None:
+            message = "{}: {}".format(message, timeout)
+        super(TimeoutError, self).__init__(message)
+
+
+class NoJobFoundError(RPCGearmanClientError):
+    def __init__(self, message=None):
+        if message is None:
             message = "No job was found"
         super(NoJobFoundError, self).__init__(message)
 
 
 INFLIGHT_POLL_TIMEOUT = 5.0
 
 
-class MCPClient:
+class MCPClient(object):
+    """MCPServer client (RPC via Gearman)."""
+
     def __init__(self):
         self.server = settings.GEARMAN_SERVER
+        self.timeout = INFLIGHT_POLL_TIMEOUT
+
+    def _rpc_sync_call(self, ability, data, timeout=INFLIGHT_POLL_TIMEOUT):
```

Comment:
I kind of think that we shouldn't abuse decorators because it affects readability. I used them in RPCServer too much and after all I realize they're not idea. There are many discussions out there about them. I'm not sure if I still have a rule of thumb, something that we should explore. For now I'll leave it like this, thanks for the suggestion though, it's worth keeping in mind in further work.